### PR TITLE
[FLINK-19585][tests] Waiting for job to run before savepointing in UnalignedCheckpointCompatibilityITCase.

### PR DESCRIPTION
**Pull request by AHeise**
## What is the purpose of the change

Fixes UnalignedCheckpointCompatibilityITCase.test:97->runAndTakeSavepoint: "Not all required tasks are currently running.", which may be caused when deploying takes longer than expected, such that stopWithSavepoint is called before everything is running.

## Brief change log

- Allow savepoints to be used in MiniCluster.
- Use JobDetailsInfo to ensure that job and all tasks are running.

## Verifying this change

Fixes test failure.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointCompatibilityITCase.java [hunk 1]
```diff
@@ -107,28 +119,55 @@ public void test() throws Exception {
...
 	private Tuple2<String, Map<String, Object>> runAndTakeSavepoint() throws Exception {
-		JobClient jobClient = submitJobInitially(env(startAligned, 0, emptyMap()));
-		Thread.sleep(FIRST_RUN_EL_COUNT * FIRST_RUN_BACKPRESSURE_MS); // wait for all tasks to run and some backpressure from sink
-		Future<Map<String, Object>> accFuture = jobClient.getAccumulators();
-		Future<String> savepointFuture = jobClient.stopWithSavepoint(false, tempFolder().toURI().toString());
-		return new Tuple2<>(savepointFuture.get(), accFuture.get());
+		return withCluster(new Configuration(), miniCluster -> {
+			JobClient jobClient = submitJobInitially(env(startAligned, 0));
+			waitForAllTaskRunning(miniCluster, jobClient.getJobID(), Deadline.fromNow(Duration.of(30, ChronoUnit.SECONDS)));
+			Thread.sleep(FIRST_RUN_BACKPRESSURE_MS); // wait for some backpressure from sink
```
Comment (rmetzger) [inline flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointCompatibilityITCase.java:125]: It's generally an anti-pattern in tests to "just sleep for a while". Can't you use the backpressure monitoring to wait until there is backpressure? (or would that take longer?)

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointCompatibilityITCase.java [hunk 2]
```diff
@@ -107,28 +119,55 @@ public void test() throws Exception {
...
-		return env.execute(dag(totalCount, false, 0, env)).getJobExecutionResult().getAllAccumulatorResults();
+		return withCluster(new Configuration(), miniCluster -> {
+			StreamExecutionEnvironment env = env(isAligned, 50);
+			final StreamGraph streamGraph = dag(totalCount, false, 0, env);
+			streamGraph.setSavepointRestoreSettings(SavepointRestoreSettings.forPath(path));
+			return env.execute(streamGraph).getJobExecutionResult().getAllAccumulatorResults();
+		});
+	}
+
+	private <T> T withCluster(Configuration configuration,
```
Comment (rmetzger) [inline flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointCompatibilityITCase.java:161]: Why are you not reusing the cluster for the different tests / jobs?

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/test/util/TestBaseUtils.java [hunk 3]
```diff
@@ -102,6 +116,24 @@ private static void verifyJvmOptions() {
...
+					JobDetailsHeaders.getInstance(),
+					params,
+					EmptyRequestBody.getInstance()).get();
+				return jobDetailsInfo.getJobStatus() == JobStatus.RUNNING &&
+					jobDetailsInfo.getJobVerticesPerState().get(ExecutionState.RUNNING) ==
+						jobDetailsInfo.getJobVertexInfos().size();
+			}, deadline, 500);
+		}
+	}
+
```
Comment (rmetzger) [inline flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/test/util/TestBaseUtils.java:136]:
```suggestion
	public static void waitForAllTaskRunning(MiniCluster miniCluster, JobID jobID, Deadline deadline) throws Exception {
		org.apache.flink.runtime.testutils.CommonTestUtils.waitUntilCondition(() -> {
			AccessExecutionGraph ec = miniCluster.getExecutionGraph(jobID).get();
			return ec.getState() == JobStatus.RUNNING && ec.getAllVertices()
				.values()
				.stream()
				.allMatch(jobVertex ->
					Arrays.stream(jobVertex.getTaskVertices()).allMatch(task -> task.getExecutionState() == ExecutionState.RUNNING)
				);
		}, deadline);
	}
```

Comment (rmetzger) [reply flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/test/util/TestBaseUtils.java:136]: Not sure about the code formatting, and I haven't tested it much. But I guess you get the idea.

**Review by rmetzger — CHANGES_REQUESTED** [review]
Thanks a lot for working on this. I left some comments inline.

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointCompatibilityITCase.java (hunk 2 above)
Comment (AHeise) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointCompatibilityITCase.java:161]: Good point. Originally, I needed to inject the configuration to restore from savepoints, but that didn't work. So I can probably leave a cluster per test. A global cluster would probably also work, I just need to make the TempFolder a class rule.

Comment (AHeise) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointCompatibilityITCase.java:161]: Good point. Originally, I needed to inject the configuration to restore from savepoints, but that didn't work. So I can probably leave a cluster per test. A global cluster would probably also work, I just need to make the TempFolder a class rule.

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointCompatibilityITCase.java (hunk 1 above)
Comment (AHeise) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointCompatibilityITCase.java:125]: The original test used sleeping for waiting until all jobs are running (hence this bug). However, whether backpressure build up or not is quite irrelevant for the semantics of the test. It's just testing that some outputs are present. I could probably wait on the output metric. 
However, it feels like LOTS of such building blocks are missing and should probably be at a more general place. (I tried to do that with `waitForAllTaskRunning` but it's easily overlooked. Ideally it should be in MiniCluster).

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/test/util/TestBaseUtils.java (hunk 3 above)
Comment (AHeise) [reply flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/test/util/TestBaseUtils.java:136]: Hm Till actually proposed to use rest client, but I can't really see why your version wouldn't work.

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointCompatibilityITCase.java (hunk 1 above)
Comment (rmetzger) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointCompatibilityITCase.java:125]: Maybe such utilities should also be made available to the user, for writing their own tests.
There are a lot of very useful utilities everywhere in our tests, but it's difficult to discover them. Ideally we have them categorized and standardized somewhere.

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/test/util/TestBaseUtils.java (hunk 3 above)
Comment (AHeise) [reply flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/test/util/TestBaseUtils.java]: With your version, I can actually move it into `flink-runtime` `CommonTestUtils`.

**Review by rmetzger — APPROVED** [review]
+1 to merge once CI has passed.

**Comment by AHeise** [comment]
E2E failures are unrelated since this PR is only affecting non-E2E tests.

**Pull request merged**